src/test/mir-opt/while-storage.rs

```diff
@@ -21,36 +21,37 @@ fn main() {
 
 // START rustc.while_loop.PreCodegen.after.mir
 // bb0: {
-//     StorageLive(_2);
 //     StorageLive(_3);
-//     _3 = _1;
-//     _2 = const get_bool(move _3) -> bb1;
+//     StorageLive(_4);
```

Comment:
Huh? This looks like something got pessimized

Reply:
So it looks like what happened was that we used to have something like this:

```
_1 = ();
```

which ConstProp didn't do anything with. The SimplifyLocals pass then removed this local. Now, ConstProp turns that into 

```
_1 = const Scalar(<ZST>) : ();
```

and SimplifyLocals no longer removes it which bumps the locals count and caused these renumberings. 

I guess there's two possible fixes here:

1. Add logic in ConstProp to check for this case and leave `()` alone.
2. Make `SimplifyLocals` smarter and remove unused ZSTs. 

Do you prefer one of those ideas? (Or something else?)

Reply:
SimplifyLocals should definitely be improved. It should ignore ZST rhs by type instead of value.

Not sure about const prop, maybe we should also skip all zsts just to keep memory and perf down

Reply:
I was wrong, `_1 = ();` was actually being optimized out by the `deaggregator` when it calls `expand_aggregate()` because the set of operands from `Rvalue::Aggregate` was empty and so the statement was being turned into a `nop`. I tried to make that code handle ZSTs but kept running into ICEs so instead, I just changed `ConstProp` to not eval these statements. I also left a FIXME. If you think that should be addressed now, let me know.

Reply:
Nah, that's a good fix. The deaggregator should work with arbitrary constants at some point, but that's not necessary in this PR.